<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: Python Coding Exercises

_Authors: Joseph Nelson (DC) _

---

The following code challenges are drawn from common exercises used in technical interviews.

Please note that there may be several ways to approach each challenge. If you get stuck, try mapping out your approach in pseudocode first. Finally, while solutions to problems like these may be found online, remember that if you copy/paste code that you can't explain, you'll be missing out on the point of the project. The only way to truly learn a new skill is through practice, trial, and error - we can only help you improve by understanding where you are having trouble.

### Challenge 1: Largest Palindrome
A palindromic number reads the same both ways. For example, 1234321 is a palindrome. The largest palindrome made from the product of two two-digit numbers is 9009 = 91 × 99. Find the largest palindrome made from the product of two three-digit numbers. Afterward, write a brief explanation walking through your code's logic in markdown.

In [18]:
#method to check if given number is palindrome
def isPalindrome(num_check):
    
    #convert to string and find middle position
    str_check = str(num_check)
    len_str = len(str_check)
    mid_pos = len_str // 2
    
    #check palindrome by checking characters from end points inwards
    for x in range(mid_pos):
        y = -x - 1
        if str_check[x] != str_check[y]:
            return False
    
    return True # returns True if all characters matches palindrome algorithm



#3-digits numbers range from 100 to 999, so product ranges from 10,000 to 998,001
for num_product in range(998001, 10000, -1):
    loop_breaker = False
    if isPalindrome(num_product):
        for x in range(999,99,-1):
            y = num_product // x
            if (num_product % x == 0) & (y <= 999):
                loop_breaker = True
                break
        if loop_breaker:
            break
            
print("The largest palindrome is {} and is the product of {} and {}".format(num_product, x, y))

The largest palindrome is 906609 and is the product of 993 and 913


### Explanation of Palindrome Logic
Function **isPalindrome** will return True if the number passed in is palindrome, returs False if otherwise.

Logic of function:
- convert input number to String
- find the middle position (if it is a odd number, then get the quotient after divide by 2)
- compare character at position 0 and -1, character at position 1 and -2, and so on...
- at any time if the 2 characters doesn't match, return False immediately without further processing
- if the entire string is compared and alll characters to be compared matches, returns True

Logic of main program:
- loop from 998001 (999*999) in decreasing order till 10001 (since 100*100 is not palindrome). Assign value num_product:
- check if the num_product is palindrome by calling isPalindrome function
- if it is palindrome, then check if it can be divisible by x = 999, 998, 997 and so on..
- if it can be fully divisible by x, check if the other divisor is a 3 digit
- if all conditions matches, then we got our largest 3-digit divisors and the palindrome.

> **note:** Initially I use another method of 2 For loops (decreasing from 999), however answer is wrong coz it hits 995 * 583 = 580085 first. After much verification, I found that 906609 = 993*913 is a larger palindrome, hence rewrite my solution.


### Challenge 2: Summation of Primes
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17. Find the sum of all the primes below 2,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [53]:
def all_primes(cap_limit):
    set_numbers = list(range(2,cap_limit))
    list_primes = []
    
    while len(set_numbers) > 0:
        prime = set_numbers[0]
        loop_limit = (cap_limit // prime )+ 1
        # prime numbers start from 2
        list_primes.append(prime) 
        for x in range(1, loop_limit):
            multiples = prime * x
            try:
                set_numbers.remove(multiples)
            except ValueError as target:
                pass
             
    return list_primes


def sum_primes(list1):    
    total = 0
    for i in range(len(list1)):
        total = total + list1[i]
    return total
        

print("The sum of all prime numbers below 2000 is {} ".format(sum_primes(all_primes(2000))))

The sum of all prime numbers below 2000 is 277050 


### Explanation of Prime Numers Logic
Reference youtube on the logic of finding Prime Numbers:
https://www.youtube.com/watch?time_continue=173&v=FBbHzy7v2Kg


2 functions are defined:
- **all_primes** will return list of prime numbers below a given numerical cap_limit 
- **sum_primes** will return the sum of a given list of numbers.

Logic of function **all_primes**:
- create a list of running numbers, `set_numbers`, starting from 2 and ending with (cap_limit - 1), as per requirement
- create a empty list `list_primes` to hold prime numbers
- Loop while `set_numbers` is not empty:
    + get the first number, prime, from `set_numbers`. Insert into `list_primes`. (for 1st cycle, value is 2)
    + get the loop_limit of second loop, which is quotient of cap_limit divide by 2, increment by 1 as For loop is not inclusive
    + the second loop will remove all the multiples of prime from `set_numbers`
    + try-except is written as the remove method will throw ValueError if value not in list

Logic of function **sum_primes**:
- create a For loop to loop through all the numbers in the list and sum them up

Logic of main program:
- print the sum of primes which are below 2000

### Challenge 3: Multiples of 3 and 5
If we list all of the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6, and 9. The sum of these multiples is 23. Find the sum of all the multiples of 3 and 5 below 1,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [63]:
def multiples_3_5(cap_limit):
    # set is used as it is unique
    natural_numbers = set()
    
    limit_3 = (cap_limit // 3) + 1
    for x in range(1,limit_3):
        multi_3 = 3 * x
        if (multi_3 < cap_limit):
            natural_numbers.add(multi_3)
    
    limit_5 = (cap_limit // 5) + 1
    for y in range(1,limit_5):
        multi_5 = 5 * y
        if (multi_5 < cap_limit):
            natural_numbers.add(multi_5) 
    return natural_numbers

def sum_numbers(set1):
    total = 0
    for i in range(len(set1)):
        total = total + set1.pop()
    return total

print("Sum of natural numbers which are multiples of 3 or 5 is {}.".format(sum_numbers(multiples_3_5(1000))))

Sum of natural numbers which are multiples of 3 or 5 is 233168.


### Explanation of Multiples of 3 and 5 Logic


2 functions are defined:
- **multiples_3_5** will return set of multiples of 3 and 5 below a given numerical cap_limit 
- **sum_numbers** will return the sum of a given set of numbers.

Logic of function **multiples_3_5**:
- create a empty set. Set is used as it is unique, so duplicate values would not be added to the set
- use for loops to add the values of multiples of 3 and multiples of 5 to the set

Logic of function **sum_numbers**:
- create a For loop to sum up all elements in the set.

Logic of main program:
- print the sum of all natural numbers which are either multiple of 3 or 5

### Challenge 4: String Compressor
Implement a method to perform basic string compression using the counts of repeated characters. (This is called run-length encoding.) For example, the string "aabcccccaaa" would become a2b1c5a3. If the “compressed” string would not become smaller than the original string, your method should return the original string. You can assume the string has only uppercase and lowercase letters (a–z). Specify whether your solution is case sensitive or case insensitive and what you would need to change to make it the other. Afterward, write a brief explanation walking through your code's logic in markdown.

In [ ]:
def checkString

In [ ]:
### Explanation of String Compressor Logic


2 functions are defined:
- **multiples_3_5** will return set of multiples of 3 and 5 below a given numerical cap_limit 
- **sum_numbers** will return the sum of a given set of numbers.

Logic of function **multiples_3_5**:
- create a empty set. Set is used as it is unique, so duplicate values would not be added to the set
- use for loops to add the values of multiples of 3 and multiples of 5 to the set

Logic of function **sum_numbers**:
- create a For loop to sum up all elements in the set.

Logic of main program:
- print the sum of all natur

### *BONUS* Challenge: FizzBuzz
Write a program that prints all of the numbers from 1 to 100. For multiples of 3, instead of the number, print "Fizz;" for multiples of 5, print "Buzz." For numbers that are multiples of both 3 and 5, print "FizzBuzz." Afterward, write a brief explanation walking through your code's logic in markdown.

In [5]:
set_numbers = list(range(1,101))
list_multi3 = list(range(3,101,3))
list_multi5 = list(range(5,101,5))
for x in set_multi5:
    index_x = x - 1
    set_numbers[index_x] = "Buzz"
    
for y in set_multi3:
    index_y = y - 1
    if y % 5 == 0:
        set_numbers[index_y] = "FizzBuzz"
    else:
        set_numbers[index_y] = "Fizz"
    
    
print(set_numbers)

[1, 2, 'Fizz', 4, 'Buzz', 'Fizz', 7, 8, 'Fizz', 'Buzz', 11, 'Fizz', 13, 14, 'FizzBuzz', 16, 17, 'Fizz', 19, 'Buzz', 'Fizz', 22, 23, 'Fizz', 'Buzz', 26, 'Fizz', 28, 29, 'FizzBuzz', 31, 32, 'Fizz', 34, 'Buzz', 'Fizz', 37, 38, 'Fizz', 'Buzz', 41, 'Fizz', 43, 44, 'FizzBuzz', 46, 47, 'Fizz', 49, 'Buzz', 'Fizz', 52, 53, 'Fizz', 'Buzz', 56, 'Fizz', 58, 59, 'FizzBuzz', 61, 62, 'Fizz', 64, 'Buzz', 'Fizz', 67, 68, 'Fizz', 'Buzz', 71, 'Fizz', 73, 74, 'FizzBuzz', 76, 77, 'Fizz', 79, 'Buzz', 'Fizz', 82, 83, 'Fizz', 'Buzz', 86, 'Fizz', 88, 89, 'FizzBuzz', 91, 92, 'Fizz', 94, 'Buzz', 'Fizz', 97, 98, 'Fizz', 'Buzz']


### Explanation of FizzBuzz Logic

Define 3 lists, one is for list of numbers from 1 to 100, another is for list of multiples of 3 below 100, and the last is for multiples of 5 not exceeding 100.

Create a For loop, for multiples of 5, update the corresponding index of list (reduce by 1 coz index starts from 0) to value "Buzz".
Repeat similar For loop logic for multiples of 3 but with value "Fizz", this time have an additional check that if the number is exactly divisible 5 (remainder 0), then update to "FizzBuzz"